src/zigbee2mqtt/zigbee2mqtt-adapter.ts

```diff
@@ -0,0 +1,261 @@
+      const raw = message.toString();
+
+      if (this.debug) {
+        console.log(`Received on ${topic}: ${raw}`);
+      }
+
+      try {
+        const json = JSON.parse(raw);
+
+        const parts = topic.split('/');
+
+        if (topic.indexOf(DEVICES_POSTFIX) > -1) {
+          this.handleDevices(client, json);
+        } else if (parts.length == 2) {
+          const id = parts[1];
+          const device = this.getDevice(id) as Zigbee2MqttDevice;
+
+          if (device) {
+            device.update(json);
+          } else {
+            console.log(`Could not find device with id ${id}`);
+          }
+        } else if (topic.indexOf(PERMIT_RESPONSE_POSTFIX) > -1) {
+          const response: Response = json;
+
+          if (response.error) {
+            console.log(`Could not enable permit join mode: ${response.error}`);
+          } else if (response.status === 'ok') {
+            if (response.data?.value) {
+              console.log('Bridge is now permitting new devices to join');
+            } else {
+              console.log('Bridge is no longer permitting new devices to join');
+            }
+          }
+        } else if (topic.indexOf(REMOVE_RESPONSE_POSTFIX) > -1) {
+          const response: Response = json;
+          const id = response.data?.id ?? 'unknown';
+
+          if (response.error) {
+            console.log(`Could not remove device ${id}: ${response.error}`);
+          } else if (response.status === 'ok') {
+            console.log(`Removed device ${id} successfully`);
+
+            const existingDevice = this.getDevice(id);
+
+            if (existingDevice) {
+              this.handleDeviceRemoved(existingDevice);
+            } else {
+              console.warn(`Could not find device with id ${id}`);
+            }
+          }
+        } else if (topic.indexOf(LOGGING_POSTFIX) > -1) {
+          const log: Log = json;
+          console.log(`Zigbee2Mqtt::${log.level}: ${log.message}`);
+        }
+      } catch (error) {
+        console.error(`Could not process message ${raw}: ${error}`);
+      }
+    });
+  }
+
+  private subscribe(topic: string): void {
+    console.log(`Subscribing to ${topic}`);
+
+    if (!this.client) {
+      console.log('No client to subscribe to');
+      return;
+    }
+
+    this.client.subscribe(topic, (err) => {
+      if (err) {
+        console.error(`Could not subscribe to ${topic}: ${err}`);
+      } else {
+        console.log(`Successfully subscribed to ${topic}`);
+      }
+    });
+  }
+
+  private handleDevices(client: mqtt.Client, deviceDefinitions: DeviceDefinition[]): void {
+    if (!Array.isArray(deviceDefinitions)) {
+      console.log(`Expected list of devices but got ${typeof deviceDefinitions}`);
+      return;
+    }
+
+    for (const deviceDefinition of deviceDefinitions) {
+      if (deviceDefinition.type == 'EndDevice' || deviceDefinition.type == 'Router') {
+        const id = deviceDefinition.friendly_name;
```

Comment:
```typescript
const id = deviceDefiniton.ieee_address;
```

That would make more sense, given that it's guaranteed to be unique, and it should work as a drop-in replacement for the API.

Reply:
done